joblib/numpy_pickle.py

```diff
@@ -1,229 +1,238 @@
-
-    This object store the Zfile filename in which
-    the data array has been persisted, and the meta information to
-    retrieve it.
-
-    The reason that we store the raw buffer data of the array and
-    the meta information, rather than array representation routine
-    (tostring) is that it enables us to use completely the strided
-    model to avoid memory copies (a and a.T store as fast). In
-    addition saving the heavy information separately can avoid
-    creating large temporary buffers when unpickling data with
-    large arrays.
-    """
-    def __init__(self, filename, init_args, state):
-        "Store the useful information for later"
-        self.filename = filename
-        self.state = state
-        self.init_args = init_args
-
-    def read(self, unpickler):
-        "Reconstruct the array from the meta-information and the z-file"
-        # Here we a simply reproducing the unpickling mechanism for numpy
-        # arrays
-        filename = os.path.join(unpickler._dirname, self.filename)
-        array = unpickler.np.core.multiarray._reconstruct(*self.init_args)
-        with open(filename, 'rb') as f:
-            data = read_zfile(f)
-        state = self.state + (data,)
-        array.__setstate__(state)
-        return array
-
+                                    self.subclass, (0,), 'b')
+            return new_array.__array_prepare__(array)
+        else:
+            return array
 
 ###############################################################################
 # Pickler classes
 
+
 class NumpyPickler(Pickler):
     """A pickler to persist of big data efficiently.
 
-        The main features of this object are:
+    The main features of this object are:
+    * persistence of numpy arrays in a single file.
 
-         * persistence of numpy arrays in separate .npy files, for which
-           I/O is fast.
+    * optional compression using GzipFile, with a special care on avoid
+    temporaries.
 
-         * optional compression using Zlib, with a special care on avoid
-           temporaries.
+    Attributes
+    ----------
+    fp: file
+        File object handle used for serializing the input object.
+    protocol: int
+        Pickle protocol used. Default is pickle.DEFAULT_PROTOCOL under
+        python 3, pickle.HIGHEST_PROTOCOL otherwise.
     """
+
     dispatch = Pickler.dispatch.copy()
 
-    def __init__(self, filename, compress=0, cache_size=10, protocol=None):
-        self._filename = filename
-        self._filenames = [filename, ]
-        self.cache_size = cache_size
-        self.compress = compress
-        if not self.compress:
-            self.file = open(filename, 'wb')
-        else:
-            self.file = BytesIO()
-        # Count the number of npy files that we have created:
-        self._npy_counter = 1
+    def __init__(self, fp, protocol=None):
+        """Constructor. Store the useful information for later."""
+        self.file = fp
```

Comment:
Could we rename file to file_handle to stay consistent with Unpickler. Maybe there are complications with backward-compatibility in which case we can live with this.